src/CloudNative.CloudEvents/JsonEventFormatter.cs

```diff
@@ -1,123 +1,117 @@
-                    {
-                        attributes["data"] = Convert.FromBase64String(keyValuePair.Value.ToString());
-                        continue;
-                    }
+                    // Handle base64 encoded binaries
+                    cloudEvent.Data = Convert.FromBase64String((string)value);
+                    continue;
                 }
-
-                switch (keyValuePair.Value.Type)
+                if (key == Data)
                 {
-                    case JTokenType.String:
-                        attributes[keyValuePair.Key] = keyValuePair.Value.ToObject<string>();
-                        break;
-                    case JTokenType.Date:
-                        // TODO: Check this is appropriate. (Should we use Timestamps instead?)
-                        attributes[keyValuePair.Key] = keyValuePair.Value.ToObject<DateTimeOffset>();
-                        break;
-                    case JTokenType.Uri:
-                        attributes[keyValuePair.Key] = keyValuePair.Value.ToObject<Uri>();
-                        break;
-                    case JTokenType.Null:
-                        attributes[keyValuePair.Key] = null;
-                        break;
-                    case JTokenType.Integer:
-                        attributes[keyValuePair.Key] = keyValuePair.Value.ToObject<int>();
-                        break;
-                    default:
-                        attributes[keyValuePair.Key] = (dynamic)keyValuePair.Value;
-                        break;
+                    // FIXME: Deserialize where appropriate.
+                    // Consider whether there are any options here to consider beyond "string" and "object".
+                    // (e.g. arrays, numbers etc).
+                    // Note: the cast to "object" is important here, otherwise the string branch is implicitly
+                    // converted back to JToken...
+                    cloudEvent.Data = value.Type == JTokenType.String ? (string) value : (object) value;
+                    continue;
                 }
+
+                var attribute = cloudEvent.GetAttribute(key);
+
+                // Set the attribute in the event, taking account of mismatches between the type in the JObject
+                // and the attribute type as best we can.
+
+                // TODO: This currently performs more conversions than it really should, in the cause of simplicity.
+                // We basically need a matrix of "attribute type vs token type" but that's rather complicated.
+
+                string attributeValue = value.Type switch
+                {
+                    JTokenType.String => (string)value,
+                    JTokenType.Date => CloudEventAttributeType.Timestamp.Format((DateTimeOffset)value),
+                    JTokenType.Uri => CloudEventAttributeType.UriReference.Format((Uri)value),
+                    JTokenType.Null => null, // TODO: Check we want to do this. It's a bit weird.
```

Comment:
Null values are not allowed per the spec, but do we want to be more permissive when decoding?